rotkehlchen/chain/ethereum/modules/weth/decoder.py

```diff
@@ -0,0 +1,180 @@
+        self.eth = A_ETH.resolve_to_crypto_asset()
+
+    def _decode_weth(
+            self,
+            tx_log: EthereumTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EthereumTxReceiptLog],
+            action_items: Optional[List[ActionItem]],
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        if tx_log.topics[0] == WETH_DEPOSIT_TOPIC:
+            return self._decode_deposit_event(
+                tx_log=tx_log,
+                transaction=transaction,
+                decoded_events=decoded_events,
+                all_logs=all_logs,
+                action_items=action_items,
+            )
+
+        if tx_log.topics[0] == WETH_WITHDRAW_TOPIC:
+            return self._decode_withdrawal_event(
+                tx_log=tx_log,
+                transaction=transaction,
+                decoded_events=decoded_events,
+                all_logs=all_logs,
+                action_items=action_items,
+            )
+
+        return None, None
+
+    def _decode_deposit_event(
+            self,
+            tx_log: EthereumTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EthereumTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[List[ActionItem]],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        depositor = hex_or_bytes_to_address(tx_log.topics[1])
+        deposited_amount_raw = hex_or_bytes_to_int(tx_log.data[:32])
+        deposited_amount = asset_normalized_value(amount=deposited_amount_raw, asset=self.eth)
+
+        in_event = None
+        # when weth is interacting with a protocol e.g. Uniswap, the protocol handles the
+        # wrapping and unwrapping of WETH
+        is_interacting_with_protocol = False
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.SPEND and
+                event.counterparty in (WETH_CONTRACT, depositor) and
+                event.balance.amount == deposited_amount and
+                event.asset == self.eth
+            ):
+                if event.counterparty == depositor:
+                    is_interacting_with_protocol = True
+                else:
+                    in_event = event
+                    event.counterparty = CPT_WETH
+                    event.event_type = HistoryEventType.DEPOSIT
+                    event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                    event.notes = f'Wrap {deposited_amount} {self.eth.symbol} as {self.weth.symbol}'  # noqa: E501
+
+        if is_interacting_with_protocol is True:
+            return None, None
+
+        out_event = HistoryBaseEntry(
+            event_identifier=transaction.tx_hash,
+            sequence_index=self.base_tools.get_next_sequence_counter(),
```

Comment:
Isn't there a transfer event to decode here?

Reply:
The transfer event is what is decoded to the `Wrap XX ETH in WETH`

Reply:
no that's not what I am saying. I mean a transfer event from the WETH token contract. WETH -> user due to depositing ETH. This would normally be autodetected by the ERC20 token decoder

Reply:
It doesn't emit such event. 
```sol
    function deposit() public payable {
        balanceOf[msg.sender] += msg.value;
        Deposit(msg.sender, msg.value);
    }
 ```

Reply:
I see. Yeah I checked the code now. It does not emit any such events in deposits or withdrawal. you can only use the events you are using here to decode.